### ISSUE TITLE
APIKeyModal release review findings 

### ISSUE THREAD

**Issue by lee-chase** [carbon-design-system/ibm-cloud-cognitive#954]
Review findings for #910 

- [ ] Scrollbar on APIKey display panel
![image](https://user-images.githubusercontent.com/15086604/123942365-fa234480-d992-11eb-80d6-bdc269c1a7a3.png)
- [ ]

**Comment by davidmenendez** [comment]
> Question: Do we need issues to track creation of other scenarios shown in the design? e.g. Edit API Key.

this appears to be a more recent additional. i'll have to go back and add a story for this.

> Rendering glitch: there is a horizontal scrollbar that appears on the APIKey display panel. You might need to turn on the option to always show scrollbars to see this, or just try side-scrolling on the key display panel to trigger the scrollbar to appear. It seems to be caused by the spaced reserved for the tooltip (and used by it, when it displays).

i'm unable to replicate this. can you explain how you were able to get this behavior?

> Story glitch: in the custom step story, checking All resources does not enable the generate button. Also, the user can still edit Which resources when All resources is checked. Checking All resources should enable the generate button. Also, either Which resources should be disabled when All resources is checked or typing something in the Which resources should uncheck All resources.

> In stories, action('submitted') should actually be action('submitted')(), in order to call the action as well as create it.

> Remove apiKeyInputId prop and use internalId approach as in ExportModal.

>  Rename apiKeyVisibility to hasApiKeyVisibilityToggle, and downloadable to hasDownloadLink, for consistency with other similar props e.g. hasCloseIcon on Tearsheet.

will fix

> TBC -- Functions inside the render function like isPrimaryButtonDisabled() and getHeader() don't just use prop values but also use stateful values like currentStep -- React question: should they therefore also be useState functions, so that React knows when the results of those functions will have changed and can decide to re-render accordingly?

i hadn't really considered this before to be honest, but yes you're right this is a great use case for `useEffect` and `useState`. i'll refactor to use those.

> Rename apiKeyVisibility to hasApiKeyVisibilityToggle, and downloadable to hasDownloadLink, for consistency with other similar props e.g. hasCloseIcon on Tearsheet.

> TBC -- Functions inside the render function like isPrimaryButtonDisabled() and getHeader() don't just use prop values but also use stateful values like currentStep -- React question: should they therefore also be useState functions, so that React knows when the results of those functions will have changed and can decide to re-render accordingly?

> TBC -- Rename createHeader and successHeader to createTitle and successTitle. We use the name "title" consistently for the main heading text on modals, e.g. on ExportModal.

> TBC -- Remove stepHeaders prop and instead include a title field in the shape for customSteps. This means that each custom step is provided as a title, content, and a valid flag.

> TBC -- nextStepButtonText and previousStepButtonText should be required if more than one custom step is provided. This can be done using our new isRequired.if, something like this nextStepButtonText: PropTypes.string.isRequired.if(({ customSteps }) => customSteps.length > 1),.

> TBC -- Commentary for the props is a bit brief and could benefit from a bit more explanation, especially as to how the props are used when custom steps are included (e.g. for props like primaryButtonText) and what/how exactly the value appears in the rendering (e.g. for props like modalLabel). Also, a couple of copy/paste issues (e.g. secondaryButtonText prop) and some spellings to clean up (e.g. "specifices" appears a few times).

will fix

**Comment by davidmenendez** [comment]
@lee-chase added a PR to address some of these issues https://github.com/carbon-design-system/ibm-cloud-cognitive/pull/985. sanity check- did you add additional items to the checklist?

**Comment by lee-chase** [comment]
@davidmenendez yes some items were added, sorry for not making it clear that @dcwarwick and I had not finished working through the component review.

**Comment by dcwarwick** [comment]
@davidmenendez a few more items have been added: first review pass now finished.

**Comment by davidmenendez** [comment]
> Design question: if the name is "required", i.e. the user cannot generate the key until they provide a name, should the name input field use the required state so that it displays the red asterisk to indicate that it is a required field?

the carbon `<TextInput />` component doesn't currently have a require prop and the designs don't specifically show anything to indicate name require. consumers should simply pass a label to indicate that the field is required.

> Rename the component APIKeyModal to GenerateAPIKeyModal, for consistency with ExportModal, CreateModal, etc, and to leave room for a possible future EditAPIKeyModal etc.

i don't agree with this. unlike `ExpressiveCard` and `ProductiveCard` the generate and edit functions don't really change the overall layout or functionality of the component. i feel like changing the name and adding an additional named export adds unnecessary complexity. it would be like if carbon didn't have a singular button component and instead had a name export for every single flavor and then the consumer had to import half a dozen buttons when one does the job.

> Design question: should the modal be a fixed height, or have a cross-step minimum height, when multiple steps are being used? The custom step story shows how the modal height jumps around from step to step, which looks a bit visually distracting and makes it hard to step quickly forwards or backwards because the buttons keep moving around. For example, the design for TagSet called for a fixed height for the "View all tags" modal (see "Hundreds of Tags" story) with scroll when needed, so that the modal doesn't change height as different search strings are entered.

hmmm this is tricky. i'll follow up with design on this.

**Comment by davidmenendez** [comment]
i've made all the updates although i'm waiting on a response to my previous comments. the sandbox also appears to be broken, so i'm unable to make one at the moment. i'll start debugging that @lee-chase @dcwarwick

**Comment by dcwarwick** [comment]
| i don't agree with this. unlike ExpressiveCard and ProductiveCard the generate and edit functions don't really change the overall layout or functionality of the component. i feel like changing the name and adding an additional named export adds unnecessary complexity. it would be like if carbon didn't have a singular button component and instead had a name export for every single flavor and then the consumer had to import half a dozen buttons when one does the job.

**Comment by dcwarwick** [comment]
> > Rename the component APIKeyModal to GenerateAPIKeyModal, for consistency with ExportModal, CreateModal, etc, and to leave room for a possible future EditAPIKeyModal etc.
> 
> i don't agree with this. unlike `ExpressiveCard` and `ProductiveCard` the generate and edit functions don't really change the overall layout or functionality of the component. i feel like changing the name and adding an additional named export adds unnecessary complexity. it would be like if carbon didn't have a singular button component and instead had a name export for every single flavor and then the consumer had to import half a dozen buttons when one does the job.

There's no "adding an additional named export": we weren't suggesting splitting a component into multiple flavors. Whether we have separate components for different scenarios or a single component should be judged on whether it makes sense to a consumer, whether the API is similar or has important differences, etc, just as we've considered for tearsheets, cards, etc. The question here is what name we should use for this component and how it addresses further scenarios and what related components might be named. The component we were reviewing only offered the generate function, and it looked like an edit function would be needed too.

It sounds like you're proposing that the same component should perform both the generate and edit API Key functions. That will be great if it makes good sense -- I haven't really looked through the edit scenario yet, but a quick glance just now does indeed suggest that it would be good to use the same component if that's feasible, as the panels and controls may be very similar, but at the same time the flow is a bit different because it ends with the 'Save' rather than the 'Generate'. I see you've added stories for the edit scenario, so we'll take a look at that and see how it all looks. But, yes, if the same component can sensibly do both functions then that's fine. Great. But it still leaves the question as to what would make a good and consistent name for that component.

**Comment by dcwarwick** [comment]
OK, one weekend later and I can't think of anything better than APIKeyModal so quite happy to leave it as it is :-) It's a bit of a swiss-army-knife component, but the basic function/layout is the same so I don't feel a strong need to make separate named wrapper components for edit/generate/multi-step/etc scenarios.

**Comment by dcwarwick** [comment]
@davidmenendez I don't think the following props are actively tested by the jest tests:
- copyIconDescription
- titles on custom steps
- downloadFileName/Type
- show/hideAPIKeyLabel
- nameHelperText
I think these can be added into the existing tests, and otherwise we're good to complete review.

**Comment by dcwarwick** [comment]
Great stuff
